os/arch/arm/src/s5j/sss/security_hal_wrapper.c

```diff
@@ -0,0 +1,1632 @@
+*/
+	return HAL_NOT_SUPPORTED;
+}
+
+int hal_rsa_sign_md(hal_rsa_mode mode, hal_data *hash, uint32_t key_idx, hal_data *sign)
+{
+	HWRAP_ENTER;
+	uint32_t ret;
+	unsigned int padding = 0;
+	struct sRSA_SIGN rsa_sign;
+	unsigned char sig_out[1024];
+
+	memset(&rsa_sign, 0, sizeof(struct sRSA_SIGN));
+
+	if (mode.rsa_a == HAL_RSASSA_PKCS1_PSS_MGF1) {
+		padding = 1;
+	}
+
+	switch (mode.hash_t) {
+	case HAL_HASH_SHA1:
+		rsa_sign.alg_type = SHA1_160 | (padding ^ 0x1);	/* 0x110X */
+		break;
+	case HAL_HASH_SHA256:
+		rsa_sign.alg_type = SHA2_256 | (padding ^ 0x1);	/* 0x230X */
+		break;
+	case HAL_HASH_SHA384:
+		rsa_sign.alg_type = SHA2_384 | (padding ^ 0x1);	/* 0x240X */
+		break;
+	case HAL_HASH_SHA512:
+		rsa_sign.alg_type = SHA2_512 | (padding ^ 0x1);	/* 0x250X */
+		break;
+	default:
+		return HAL_NOT_SUPPORTED;
+	}
+
+	rsa_sign.signature = sig_out;
+	rsa_sign.signature_byte_len = 1024;
+	rsa_sign.salt_byte_len = mode.salt_byte_len;
+
+	ISP_CHECKBUSY();
+	ret = isp_rsa_sign_md_securekey(&rsa_sign, hash->data, hash->data_len, key_idx);
+	if (ret != 0) {
+		isp_clear(0);
+		printf("ISP failed (%zu)\n", ret);
+		return HAL_FAIL;
+	}
+
+	if (rsa_sign.signature == NULL) {
+		return HAL_FAIL;
+	}
+
+	HAL_COPY_DATA(sign, rsa_sign.signature, rsa_sign.signature_byte_len);
+
+	return HAL_SUCCESS;
+}
+
+int hal_rsa_verify_md(hal_rsa_mode mode, hal_data *hash, hal_data *sign, uint32_t key_idx)
+{
+	HWRAP_ENTER;
+	uint32_t ret;
+	unsigned int padding = 0;
+	struct sRSA_SIGN rsa_sign;
+
+	memset(&rsa_sign, 0, sizeof(struct sRSA_SIGN));
+
+	rsa_sign.signature = (unsigned char *)sign->data;
+	rsa_sign.signature_byte_len = sign->data_len;
+	rsa_sign.salt_byte_len = mode.salt_byte_len;
+
+	if (mode.rsa_a == HAL_RSASSA_PKCS1_PSS_MGF1) {
+		padding = 1;
+	}
+
+	switch (mode.hash_t) {
+	case HAL_HASH_SHA1:
+		rsa_sign.alg_type = SHA1_160 | (padding ^ 0x1);	/* 0x110X */
+		break;
+	case HAL_HASH_SHA256:
+		rsa_sign.alg_type = SHA2_256 | (padding ^ 0x1);	/* 0x230X */
+		break;
+	case HAL_HASH_SHA384:
+		rsa_sign.alg_type = SHA2_384 | (padding ^ 0x1);	/* 0x240X */
+		break;
+	case HAL_HASH_SHA512:
+		rsa_sign.alg_type = SHA2_512 | (padding ^ 0x1);	/* 0x250X */
+		break;
+	default:
+		return HAL_NOT_SUPPORTED;
+	}
+
+	ISP_CHECKBUSY();
+	ret = isp_rsa_verify_md_securekey(&rsa_sign, hash->data, hash->data_len, key_idx);
+	if (ret != 0) {
+		isp_clear(0);
+		printf("ISP failed (%zu)\n", ret);
+		return HAL_FAIL;
+	}
+
+	return HAL_SUCCESS;
+}
+
+int hal_ecdsa_sign_md(hal_data *hash, uint32_t key_idx, hal_ecdsa_mode *mode, hal_data *sign)
+{
+	HWRAP_ENTER;
+	uint32_t ret;
+	unsigned char s_buf[HAL_MAX_ECP_KEY_SIZE_ALT];
+	unsigned char r_buf[HAL_MAX_ECP_KEY_SIZE_ALT];
+	unsigned char sign_buf[HAL_MAX_BUF_SIZE];
+
+	struct sECC_SIGN ecc_sign;
+	memset(&ecc_sign, 0, sizeof(struct sECC_SIGN));
+
+	ecc_sign.s = s_buf;
+	ecc_sign.s_byte_len = HAL_MAX_ECP_KEY_SIZE_ALT;
+	ecc_sign.r = r_buf;
+	ecc_sign.r_byte_len = HAL_MAX_ECP_KEY_SIZE_ALT;
+	
+	switch (mode->hash_t) {
```

Comment:
I think you can make function to handle this switch state to reduce code line.

Reply:
Thanks for the comment. Unfortunately, specific conditions within the switch statement differ slightly for each case. In addition, curve types for specific modules supported by the SE may vary.